### ISSUE TITLE
Install takes very long time when ran by pipeline's playbook.

### ISSUE THREAD

**Issue by trlvsk** [pyenv/pyenv#3099]
Hello everyone.

AWS instance :t4gmedium
Pyenv Version: v2.4.17

I'm facing anoying optimisation issue with Pyenv on my pipeline. 
In short pipeline creates AMI with packages for EC2 service running pyenv+poetry and couple more things.
All of the tasks are automated and done wtih Packer/TF/Ansible.

In this case Pyenv and then the pyenv version is installed using ansible (code below).
The issue is when it's ran on the pipeline (bamboo agents) just the task that runs **pyenv install 3.9.x** takes 20-30 minutes which is not acceptable for my use case.

On the other hand when i disabled this specific playbook within the pipeline and then I ran exactly the same playbook on the other t4gmedium instance (so same packages same config same everything realy) manualy the whole playbook took about 4 to 5 minutes so the offshoot is very clear here and as of now I'm really out of ideas why is that.

It's worth noting that in each of the scenarios both very long pipeline one and the shorter manual one the outcome is identical.
My code is completely operational, poetry works just fine.
The matter here is only the build time.
I've also seen quite similar issue to this but in those cases users were using very tiny instances which is not the case here.
I've also monitored the builder instance and it's not having resource nor network issues at all.

In the logs from bamboo I can see very very long pause that I cannot figure out. + permissions for shims.

**Logs**

```
build	30-Oct-2024 08:56:15	[0;32m    amazon-ebs.service_ami: TASK [Install Python 3.9.12 using pyenv] ***************************************[0m
build	30-Oct-2024 08:56:15	[0;32m    amazon-ebs.service_ami: task path: /data/bamboo-agent/xml-data/build-dir/SMOPS-TESTAMIJUNEENDEND-BUILDAMIARM64AL2023/service/ami/al2023/post-install-tasks2.yaml:6[0m
build	30-Oct-2024 09:21:37	[0;32m    amazon-ebs.service_ami: <127.0.0.1> ESTABLISH SSH CONNECTION FOR USER: ec2-user[0m
build	30-Oct-2024 09:21:37	[0;32m    amazon-ebs.service_ami: <127.0.0.1> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=37265 -o 'IdentityFile="/tmp/ansible-key4056627000"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="ec2-user"' -o ConnectTimeout=10 -o IdentitiesOnly=yes -o 'ControlPath="/root/.ansible/cp/5a4980148c"' 127.0.0.1 '/bin/sh -c '"'"'echo ~ec2-user && sleep 0'"'"''[0m
build	30-Oct-2024 09:21:37	[0;32m    amazon-ebs.service_ami: <127.0.0.1> (0, b'/home/ec2-user\n', b'mkdir: cannot create directory \xe2\x80\x98/opt/.pyenv/shims\xe2\x80\x99: Permission denied\nmkdir: cannot create directory \xe2\x80\x98/opt/.pyenv/versions\xe2\x80\x99: Permission denied\n').........
```

**Playbook**
```
- name: Ensure /opt/pyenv directory exists
  become: true
  file:
    path: /opt/.pyenv
    state: directory
    mode: '0755'

- name: Clone pyenv repository into /opt
  become: true
  git:
    repo: "https://github.com/pyenv/pyenv.git"
    dest: /opt/.pyenv
    version: v2.4.17

- name: Set up pyenv globally for all users
  become: true
  copy:
    dest: /etc/profile.d/pyenv.sh
    content: |
      export PYENV_ROOT=/opt/.pyenv
      export PATH=$PYENV_ROOT/bin:$PATH
      if command -v pyenv 1>/dev/null 2>&1; then
        eval "$(pyenv init --path)"
        eval "$(pyenv init -)"
      fi
    mode: '0755'

- name: Install Python {{python_version}} using pyenv
  become: true
  shell: |
    pyenv install -v {{python_version}} 
  args:
    executable: /bin/bash
  environment:
    PYENV_ROOT: "/opt/.pyenv"
    PATH: "/opt/.pyenv/bin:/usr/local/bin:/usr/bin:/bin"

- name: Set global of pyenv to version {{python_version}}
  become: true
  shell: |
    pyenv global {{python_version}}
  args:
    executable: /bin/bash
  environment:
    PYENV_ROOT: "/opt/.pyenv"
    PATH: "/opt/.pyenv/bin:/usr/local/bin:/usr/bin:/bin"
```
    
Should you need any more details please let me know.
    
Thanks in advance,
Marcin

**Comment by native-api** [comment]
I see there's a permission issue (and you probably see that s well). Cannot say anything about the delay, that seems unrelated to Pyenv, the hoster's support is the one who may be able to help.

**Comment by trlvsk** [comment]
Hello again,

We've managed to resolve the issue, although the solution was far from logical or intuitive.
The problem was rather complex to debug, so for future reference, I’ll leave the fix below.

In general, pyenv installation is quite CPU-intensive but even on AWS instances as powerful as t4g.medium and t4g.large processes of installation were being throttled for unreasonable amount of time.
Presumably, as it's our guess, it was waiting for other processes’ load to be low enough to start this one as it was expecting burst of CPU utilization form pyenv install.

To resolve this, we enabled unlimited burstable performance instances in the EC2 configuration with Packer.
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/burstable-performance-instances-unlimited-mode-concepts.html
We are yet to monitor if and how severly will it affect the costs of the build instances but for now it's the only viable solution.

I still believe this should not be required, as these instances are beefy enough to handle the load. Moreover the same process took approximately 4–5 minutes when run manually on the instance.
In other words, it’s certainly a bug we’ll flag to AWS, but I doubt it will be addressed promptly. This is why I suggest the above workaround if you encounter a similar issue.

Regards,
Marcin
